### ISSUE TITLE
Changes in node_modules are not reflected in a monorepo

### ISSUE THREAD

**Issue by AnatoleLucet** [sveltejs/language-tools#1415]
**Describe the bug**
In a pnpm monorepo, a change made in a Svelte file in a package will not be reflected in an app using this package until the LSP is restarted.

**To Reproduce**
Clone: https://github.com/AnatoleLucet/svelte-lsp-monorepo-issue
and install the deps with pnpm.

Both the `ui` package and the `web` app are barebone SvelteKit projects.

The `ui` package exports the `Hello.svelte` component, which has a prop named `word` of type `string`.
The `web` app is importing the `ui` package and using the `Hello.svelte` component in `src/routes/index.svelte`.
If I change `word`'s type to a `number`, the LSP will not reflect this change in the `web` app. I need to restart the LSP to get the expected `Type 'string' is not assignable to type 'number'.`.

**Expected behavior**
Reflect the changes when another package is changed (as the typescript LSP is doing)

**Screenshots**
![](https://user-images.githubusercontent.com/35486736/158775292-c80a8a5c-eca2-4053-bc42-f3d2e695571b.gif)

**System (please complete the following information):**
 - OS: Linux Manjaro 21.2.3
 - IDE: NeoVim 0.6.1 with the lsp-config plugin
 - Plugin/Package: svelte-language-server 0.14.23

**Additional context**
I'm not sure why, but I do not see the issue in VsCode.
You can check out the NeoVim Svelte lsp-config [here](https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/server_configurations/svelte.lua#L3-L4).

**Comment by dummdidumm** [comment]
Sounds related to #1301 (or maybe even a duplicate of it)

**Comment by AnatoleLucet** [comment]
Indeed it looks like a dup. Feel free to close this issue

**Comment by jasonlyu123** [comment]
But why is it only happening in neovim? I can't quite think of a reason why this would differ from the LSP clients. From my debugging, It does get resolved to the real path in the server. 

In VScode the problem is just the diagnostic not being updated in real-time. That's because we only update the diagnostic in other files when the file saves for better performance. But hover would still show the updated type. Not familiar with neovim. Is the file being saved to the fs in your screen recording? Also is this a split-screen of the monorepo root? Or are you opening two instances for each workspace?

**Comment by AnatoleLucet** [comment]
> Is the file being saved to the fs in your screen recording?

Yes. It's barely visible, but I do save the file (FYI, it's when I type `:w` in the very bottom left corner of the screen).

> Also is this a split-screen of the monorepo root? Or are you opening two instances for each workspace?

This is a single instance of neovim - at the root of the monorepo - with two buffers (files) open side-by-side.

Is there anything I can do to help debug this? Maybe spinning up the server in local an adding logs that you might need?

**Comment by jasonlyu123** [comment]
I am also wondering if this problem is os specific? Can you try running this script in the monorepo root and see if it's resolved to the real path?
```js
const fs = require('fs');

console.log(fs.realpathSync.native('apps/web/node_modules/ui/src/lib/Hello.svelte'));
```

Also, if you're interested in debugging this. You can follow [this instruction](https://nodejs.org/en/docs/guides/debugging-getting-started/) to start the server under inspect flag. And use chrome or edge to debug. if it hit this line with a symlink path then it's the same problem with 1301. It's just for some reason not resolved like what I saw in vscode. 
https://github.com/sveltejs/language-tools/blob/5d7f676a8b6649e5ab5eab0ccedcd4733e65450f/packages/language-server/src/plugins/typescript/SnapshotManager.ts#L108

**Comment by AnatoleLucet** [comment]
`fs.realpathSync` seems to be able to resolve the correct path

![](https://user-images.githubusercontent.com/35486736/158988337-1f47caa1-a47b-445e-9927-191369392158.png)

How can I use the language server codebase locally to debug? I see there's a `watch` script in the `package.json`. Should I run this and install the local package globally, or is there a better way?

**Comment by jasonlyu123** [comment]
We usually just use vscode to launch the extension together with the server. And use vscode to debug. But since you might want to debug with neovim. You'll have to add a --inspect flag to your server command. And attach the debugger to the node process. A chrome browser should be the easiest. Or you could also clone our repo and use our debugger config (attach) with a `--inspect=6009` flag in your server command config to debug on vscode. 

I have to find some time to figure out how to set up a neovim with LSP, either on windows or WSL. It probably is the second most used LSP client of svelte language server. but I don't know how to set it up to debug it.

**Comment by AnatoleLucet** [comment]
I took a bit of time to debug.
I'm not familiar with the codebase, so I'm not sure where to look at, but I did check in the `SnapshotManager`, and the `fileName` seems to be correct (`/tmp/svelte-lsp-monorepo-issue/packages/ui/src/lib/Hello.svelte`).

**Comment by jasonlyu123** [comment]
Finally got neovim to work for me. I tried using `vsplit` to open these two files in split-screen. And there're two instances of the language-server process running.

![圖片](https://user-images.githubusercontent.com/36730922/161186185-d00564d7-1ee8-469c-966c-627a0be72998.png)
![圖片](https://user-images.githubusercontent.com/36730922/161186057-4756e853-7978-433b-9b83-4da5cb25b79a.png)

From this [config](https://github.com/neovim/nvim-lspconfig/blob/3c53bf17e5478d44111027ca1a78190edff13445/lua/lspconfig/server_configurations/svelte.lua#L14), it seems like `neovim/lspconfig` would walk through the directory to find a `.git` or `package.json` and open the LSP server at the path. In this case, it isn't related to #1301. And the solution we discussed in that issue won't work for this case. The only solution I can think of is to open a dedicated file watcher for a single file like how the tsserver does.

**Comment by ryoppippi** [comment]
I'm in the same problem with my sveltekit project.
I'm using `typescript-svelte-plugin`.
Although typescript files (like +page.server.ts) works realtime, diagnotics of `.svelte` file does not updated.

**Comment by jasonlyu123** [comment]
@ryoppippi This seems unrelated to this issue. This issue is neovim creating multiple instances in monorepo. If you are indeed having the problem, you can try to override the root_dir config to `root_pattern(".git")` or any file in the root directory. 

https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#svelte

**Comment by ryoppippi** [comment]
OK this is slightly different from Monorepo problem, but the behavior is the same.
I'm doing `root_pattern(  {"package.json", "tsconfig.json", "tsconfig.jsonc", "node_modules"})`, but it seems not working

**Comment by jasonlyu123** [comment]
Then it's unrelated.

**Comment by ryoppippi** [comment]
Do I need to create new issue for that?
